src/audio/rtnr/rtnr.c

```diff
@@ -0,0 +1,554 @@
+		if (source->source->state == dev->state) {
+			if (num_sources < RTNR_MAX_SOURCES) {
+				sources[num_sources] = source;
+				sources_stream[num_sources] = &source->stream;
+				num_sources++;
+			} else {
+				buffer_unlock(source, flags);
+			}
+		} else {
+			buffer_unlock(source, flags);
+		}
+	}
+
+	/* check if there are any sources active */
+	if (num_sources == 0 || num_sources > 2)
+		return -EINVAL;
+
+	#if RTNR_ENABLE_PROCESSING
+	/* put dummy data into output queue*/
+	int nch = sources_stream[0]->channels;
+	int copy_size = BLOCKLENGTH*nch*sizeof(int32_t);
+
+	RTKMA_API_First_Copy((void *)&cd->rtk_agl, copy_size);
+	#endif // RTNR_ENABLE_PROCESSING
+
+	sink = list_first_item(&dev->bsink_list, struct comp_buffer, source_list);
+	buffer_lock(sink, &flags);
+	frames = audio_stream_avail_frames(sources_stream[0], &sink->stream);
+	buffer_unlock(sources[0], flags);
+
+	/* Process integer multiple of RTNR internal block length */
+	num_blocks = frames / RTNR_BLK_LENGTH;
+	frames = num_blocks * RTNR_BLK_LENGTH;
+
+	if (cd->count) {
+		cd->count--;
+	} else {
+		for (int i = 0; i < num_sources; ++i)
+			comp_info(dev, "rtnr_copy() sources[%d]->id: %d, frames = %d",
+				i, sources[i]->id, frames);
+		cd->count = 249; // about once per second when 16kHz
+	}
+
+	if (frames) {
+		/* Run processing function */
+		cd->rtnr_func(dev, sources_stream, &sink->stream, frames);
+
+		/* Track consume and produce */
+		source_bytes = frames * audio_stream_frame_bytes(sources_stream[0]);
+		sink_bytes = frames * audio_stream_frame_bytes(&sink->stream);
+		comp_update_buffer_consume(sources[0], source_bytes);
+		comp_update_buffer_produce(sink, sink_bytes);
+	}
+
+	#if RTNR_ENABLE_PROCESSING
+	/*real process function of rtk ma*/
+	copy_size = BLOCKLENGTH*MicNum*sizeof(int32_t);
+	RTKMA_API_Process((void *)&cd->rtk_agl, copy_size, 0);
+	#endif // RTNR_ENABLE_PROCESSING
+
+	return 0;
+}
+
+static int rtnr_prepare(struct comp_dev *dev)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+	struct comp_buffer *sinkb;
+	int ret;
+
+	comp_info(dev, "rtnr_prepare()");
+
+	ret = comp_set_state(dev, COMP_TRIGGER_PREPARE);
+	if (ret < 0)
+		return ret;
+
+	if (ret == COMP_STATUS_STATE_ALREADY_SET)
+		return PPL_STATUS_PATH_STOP;
+
+	/* Get sink data format */
+	sinkb = list_first_item(&dev->bsink_list, struct comp_buffer, source_list);
+	cd->sink_format = sinkb->stream.frame_fmt;
+
+	/* Check source and sink PCM format and get copy function */
+	comp_info(dev, "rtnr_prepare(), sink_format=%d", cd->sink_format);
+	cd->rtnr_func = rtnr_find_func(cd->sink_format);
+	if (!cd->rtnr_func) {
+		comp_err(dev, "rtnr_prepare(): No suitable processing function found.");
+		goto err;
+	}
+
+#if RTNR_ENABLE_PROCESSING
+	/* Clear in/out buffers */
+	RTKMA_API_Prepare((void *)&cd->rtk_agl);
+#endif // RTNR_ENABLE_PROCESSING
+
+	/* Initialize RTNR */
+	return 0;
+
+err:
+	comp_set_state(dev, COMP_TRIGGER_RESET);
+	return ret;
+}
+
+static int rtnr_reset(struct comp_dev *dev)
```

Comment:
do we need to reset any state here? I have a hard time believeing there isn't any state from the previous stream to reset

Reply:
Please reset the content of comp_data here.

Reply:
The reset of the content of comp_data is done in RTKMA_API_Prepare.

Reply:
sink_format should still be cleared here